src/core/trulens/experimental/otel_tracing/core/instrument.py

```diff
@@ -81,3 +92,57 @@ def wrapper(*args, **kwargs):
         return wrapper
 
     return inner_decorator
+
+
+class App(core_app.App):
+    # For use as a context manager.
+    def __enter__(self):
+        logging.debug("Entering the OTEL app context.")
+
+        # Note: This is not the same as the record_id in the core app since the OTEL
+        # tracing is currently separate from the old records behavior
+        otel_record_id = str(uuid.uuid4())
+
+        tracer = trace.get_tracer_provider().get_tracer(TRULENS_SERVICE_NAME)
+
+        # Calling set_baggage does not actually add the baggage to the current context, but returns a new one
+        # To avoid issues with remembering to add/remove the baggage, we attach it to the runtime context.
+        self.token = context_api.attach(
+            set_baggage(SpanAttributes.RECORD_ID, otel_record_id)
+        )
+
+        # Use start_as_current_span as a context manager
+        self.span_context = tracer.start_as_current_span("root")
+        root_span = self.span_context.__enter__()
+
+        logger.debug(str(get_baggage(SpanAttributes.RECORD_ID)))
+
+        root_span.set_attribute("kind", "SPAN_KIND_TRULENS")
+        root_span.set_attribute("name", "root")
+        root_span.set_attribute(
+            SpanAttributes.SPAN_TYPE, SpanAttributes.SpanType.RECORD_ROOT
+        )
+        root_span.set_attribute(
+            SpanAttributes.RECORD_ROOT.APP_NAME, self.app_name
+        )
+        root_span.set_attribute(
+            SpanAttributes.RECORD_ROOT.APP_VERSION, self.app_version
+        )
+        root_span.set_attribute(SpanAttributes.RECORD_ROOT.APP_ID, self.app_id)
+        root_span.set_attribute(
+            SpanAttributes.RECORD_ROOT.RECORD_ID, otel_record_id
+        )
```

Comment:
the semantic convention that Piotr laid out also has the `MAIN_INPUT`/`MAIN_OUTPUT`/`MAIN_ERROR`, but I don't see how we can get that even during `__exit__`, so I'm inclined to say we can remove them. Though that raises the question of, how will the UI know what the input and output are to display. It's too close to the finish line of 2024 that I don't want to investigate how this is determined currently but can next year haha.

There's also `TOTAL_COST` but I'm not as worried about that.

Reply:
yeah we should chat about this - I think one idea I had while typing this up was to omit the record_root span type, and instead just store the relevant attributes (app name/version/id/record id) in the baggage. record root can just be identified as the span with no parent

Reply:
We still need the root span type though to denote when we start a record though since if you create a otel trace before you start calling your app a bunch of times they'll all be part of the same trace but we want them to be separate records.

But more importantly, I'm confused, how does this help us determine the `MAIN_INPUT`/`MAIN_OUTPUT`/`MAIN_ERROR`?